Some before-and-after examples.

Before:
```
acoghlan@TerminalMist:~/devel/cpython$ python3 -m inspect sys --details
Can't get info for builtin modules.
```

After:
```
$ ./python -m inspect sys --details
Target: sys
Origin: built-in
Source: None
Cached: None
Loader: <class '_frozen_importlib.BuiltinImporter'>
```

Before:
```
$ python3 -m inspect _pickle --details
Target: _pickle
Origin: None
Traceback (most recent call last):
  File "<frozen runpy>", line 198, in _run_module_as_main
  File "<frozen runpy>", line 88, in _run_code
  File "/usr/lib64/python3.13/inspect.py", line 3460, in <module>
    _main()
    ~~~~~^^
  File "/usr/lib64/python3.13/inspect.py", line 3441, in _main
    print('Cached: {}'.format(module.__cached__))
                              ^^^^^^^^^^^^^^^^^
AttributeError: module '_pickle' has no attribute '__cached__'
```

After:
```
$ ./python -m inspect _pickle --details
Target: _pickle
Origin: /home/acoghlan/devel/cpython/build/lib.linux-x86_64-3.15/_pickle.cpython-315-x86_64-linux-gnu.so
Source: None
Cached: None
Loader: <_frozen_importlib_external.ExtensionFileLoader object at 0x7e9d38f2a2c0>
```

Before:
```
$ python3 -m inspect os:path.join --details
Target: os:path.join
Origin: /usr/lib64/python3.13/os.py
Traceback (most recent call last):
  File "<frozen runpy>", line 198, in _run_module_as_main
  File "<frozen runpy>", line 88, in _run_code
  File "/usr/lib64/python3.13/inspect.py", line 3460, in <module>
    _main()
    ~~~~~^^
  File "/usr/lib64/python3.13/inspect.py", line 3441, in _main
    print('Cached: {}'.format(module.__cached__))
                              ^^^^^^^^^^^^^^^^^
AttributeError: module 'os' has no attribute '__cached__'
```

After:
```
$ ./python -m inspect os:path.join --details
Target: posixpath:join (looked up as "os:path.join")
Origin: frozen
Source: /home/acoghlan/devel/cpython/Lib/posixpath.py
Cached: None
Line: 72
```

--- a/Lib/test/test_inspect/test_inspect.py
+++ b/Lib/test/test_inspect/test_inspect.py
@@ -6540,10 +6540,16 @@ class TestModuleCLI(unittest.TestCase):
         self.assertEqual(lines, [self.BUILTIN_ERROR])
 
     def test_error_extension(self):
+        module_name = "_testcapi"
+        if module_name in sys.builtin_module_names:
+            # WASI test environment has even _testcapi as a builtin module
+            expected_error = self.BUILTIN_ERROR
+        else:
+            expected_error = self.NO_SOURCE_ERROR
         _, out, err = assert_python_failure('-m', 'inspect',
-                                            '_testcapi')
+                                            module_name)
         lines = err.decode().splitlines()
-        self.assertEqual(lines, [self.NO_SOURCE_ERROR])
+        self.assertEqual(lines, [expected_error])
 
     def test_error_data(self):
         _, out, err = assert_python_failure('-m', 'inspect',
@@ -6714,8 +6720,12 @@ class TestModuleCLI(unittest.TestCase):
         self.assertEqual(inspect.findsource(target)[1], details["lineno"])
 
     def test_get_cli_details_for_extension_module(self):
-        expected_error = self.NO_SOURCE_ERROR
         module_name = "_testcapi"
+        if module_name in sys.builtin_module_names:
+            # WASI test environment has even _testcapi as a builtin module
+            expected_error = self.BUILTIN_ERROR
+        else:
+            expected_error = self.NO_SOURCE_ERROR
         module = importlib.import_module(module_name)
         details = inspect._get_details_for_cli(module, module_name, module)
         self._check_details(module, details, {"loader"}, error=expected_error)

--- a/Lib/test/test_inspect/test_inspect.py
+++ b/Lib/test/test_inspect/test_inspect.py
@@ -6541,7 +6541,7 @@ class TestModuleCLI(unittest.TestCase):
 
     def test_error_extension(self):
         _, out, err = assert_python_failure('-m', 'inspect',
-                                            '_pickle')
+                                            '_testcapi')
         lines = err.decode().splitlines()
         self.assertEqual(lines, [self.NO_SOURCE_ERROR])
 
@@ -6638,14 +6638,14 @@ class TestModuleCLI(unittest.TestCase):
         self.assertEqual(output_lines, expected_lines)
         self.assertEqual(err, b'')
 
-    def _check_details(self, module, details, other_expected_keys=(), *, error=None, alias=None):
+    def _check_details(self, module, details, other_expected_keys=(), *, alias=None, error=None):
         expected_keys = {"target", "origin", "source", "cached"}
-        if error is not None:
-            expected_keys.add("error")
-        if alias is not None:
-            expected_keys.add("alias")
         if other_expected_keys:
             expected_keys |= other_expected_keys
+        if alias is not None:
+            expected_keys.add("alias")
+        if error is not None:
+            expected_keys.add("error")
         self.assertEqual(set(details.keys()), expected_keys)
         self.assertEqual(module.__spec__.origin, details["origin"])
         try:
@@ -6664,7 +6664,8 @@ class TestModuleCLI(unittest.TestCase):
         if alias is not None:
             self.assertEqual(details["alias"], alias)
             self.assertNotEqual(details["target"], alias)
-
+        if error is not None:
+            self.assertEqual(details["error"], error)
 
     def test_get_cli_details_for_source_module(self):
         module_name = "inspect"
@@ -6689,7 +6690,7 @@ class TestModuleCLI(unittest.TestCase):
         self.assertEqual(inspect.findsource(target)[1], details["lineno"])
 
     def test_get_cli_details_for_builtin_module(self):
-        expected_error = "No source code available for builtin module"
+        expected_error = self.BUILTIN_ERROR
         module_name = "sys"
         module = importlib.import_module(module_name)
         details = inspect._get_details_for_cli(module, module_name, module)
@@ -6713,12 +6714,12 @@ class TestModuleCLI(unittest.TestCase):
         self.assertEqual(inspect.findsource(target)[1], details["lineno"])
 
     def test_get_cli_details_for_extension_module(self):
-        expected_error = "No source code available for this module"
-        module_name = "_pickle"
+        expected_error = self.NO_SOURCE_ERROR
+        module_name = "_testcapi"
         module = importlib.import_module(module_name)
         details = inspect._get_details_for_cli(module, module_name, module)
         self._check_details(module, details, {"loader"}, error=expected_error)
-        target = module.load
+        target = module.fatal_error
         nominal_target = f"{module_name}:{target.__qualname__}"
         details = inspect._get_details_for_cli(module, nominal_target, target)
         self._check_details(module, details, error=expected_error)

--- a/Lib/test/test_inspect/test_inspect.py
+++ b/Lib/test/test_inspect/test_inspect.py
@@ -6545,9 +6545,6 @@ class TestModuleCLI(unittest.TestCase):
         lines = err.decode().splitlines()
         self.assertEqual(lines, [self.NO_SOURCE_ERROR])
 
-    # End-to-end frozen module test is intentionally omitted, as whether that
-    # works without any errors depends on if the source file is still present
-
     def test_error_data(self):
         _, out, err = assert_python_failure('-m', 'inspect',
                                             'importlib.machinery:SOURCE_SUFFIXES')
@@ -6617,6 +6614,7 @@ class TestModuleCLI(unittest.TestCase):
         self.assertEqual(output_lines, expected_lines)
         self.assertEqual(err, b'')
 
+    @unittest.skipIf(not os.path.exists(os.path.__file__), "Needs frozen source file")
     def test_details_option_with_aliased_target(self):
         # Also an end-to-end test of successful non-module lookups
         module = importlib.import_module("os.path")

